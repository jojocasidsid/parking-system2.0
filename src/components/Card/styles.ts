import styled from '@emotion/styled'

export const CardWrapper = styled('div')`
	display: flex;
	justify-content: flex-start;
	align-items: center;
	padding: 20px;
	flex-direction: column;
	border-radius: 30px;
	border: 2px solid ${({ theme }) => theme.palette.secondary.main};
`
export const StyledSlotHeader = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`
