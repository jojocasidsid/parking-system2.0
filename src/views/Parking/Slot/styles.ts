import styled from '@emotion/styled'
import DirectionsCarIcon from '@mui/icons-material/DirectionsCar'

export const StyledSlot = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`

export const StyledImage = styled(DirectionsCarIcon)`
	font-size: 45px;
`

export const StyledBody = styled('div')`
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;
	text-align: center;
`
