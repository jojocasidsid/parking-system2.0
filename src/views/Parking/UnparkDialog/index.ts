import UnparkDialog from './UnparkDialog'

export default UnparkDialog
