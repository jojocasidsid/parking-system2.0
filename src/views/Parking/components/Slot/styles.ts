import styled from '@emotion/styled'
import DirectionsCarIcon from '@mui/icons-material/DirectionsCar'

export const ThumbnailWrapper = styled('div')`
	display: flex;
	justify-content: flex-start;
	align-items: center;
	padding: 20px;
	flex-direction: column;
	border-radius: 30px;
	border: 2px solid ${({ theme }) => theme.palette.secondary.main};
`
export const StyledSlot = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`

export const StyledImage = styled(DirectionsCarIcon)`
	font-size: 45px;
`

export const ActionWrapper = styled('div')`
	& > button {
		margin: 20px;
	}
`
