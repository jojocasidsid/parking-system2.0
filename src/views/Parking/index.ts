import Parking from './Parking'

export default Parking
