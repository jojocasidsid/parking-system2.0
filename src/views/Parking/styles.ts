import styled from '@emotion/styled'

export const StyledRoot = styled('div')`
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;
	padding: 20px;
`

export const ParkingWrapper = styled('div')`
	display: flex;
	justify-content: center;
	align-items: center;
	padding: 20px;
`

export const SlotsWrapper = styled('div')`
	display: grid;
	grid-template-columns: repeat(5, 1fr);
	grid-template-rows: repeat(5, 1fr);
	grid-gap: 10px;
`

export const ActionWrapper = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
	padding: 20px;
`
