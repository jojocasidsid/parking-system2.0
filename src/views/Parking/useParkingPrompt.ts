import { useState } from 'react'

// packages\
import _ from 'lodash'
import moment from 'moment'
import { useSnackbar } from 'notistack'

// slots
import { SlotsApi, EarningsApi } from 'apis'

// helpers
import {
	computeTransaction,
	getTimeDifference,
	continousRateCalculation,
	snackbarMesages,
} from 'helpers'

const useParkingPrompt = (
	slotRefetch: () => void,
	earningsRefetch: () => void
) => {
	const { enqueueSnackbar } = useSnackbar()

	const [openLeave, setOpenLeave] = useState(false)
	const [awaitingResponse, setAwaitingResponse] = useState(false)
	const [leaveTargetId, setLeaveTargetId] = useState(0)

	const [openCancelModal, setOpenCancelModal] = useState(false)
	const [cancelTargetId, setCancelTargetId] = useState(0)

	const openModalReservation = (id: number) => {
		setOpenCancelModal(true)
		setCancelTargetId(id)
	}

	const closeModalReservation = () => {
		if (!awaitingResponse) {
			setOpenCancelModal(false)
			setCancelTargetId(0)
		}
	}

	const openModalLeave = (id: number) => {
		setOpenLeave(true)
		setLeaveTargetId(id)
	}

	const closeModalLeave = () => {
		if (!awaitingResponse) {
			setOpenLeave(false)
			setLeaveTargetId(0)
		}
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const handleError = (error: any) => {
		const errorMessage = _.get(error, 'message', '')
		if (errorMessage) {
			enqueueSnackbar(errorMessage, {
				variant: 'error',
			})
		} else {
			enqueueSnackbar('Something went wrong', {
				variant: 'error',
			})
		}
	}

	const onConfirmCancelReservation = () => {
		setAwaitingResponse(true)
		SlotsApi.leaveSlot(cancelTargetId)
			.then(() => {
				closeModalReservation()
				enqueueSnackbar(snackbarMesages.deleteReservation, {
					variant: 'success',
				})
				slotRefetch()
			})
			.catch((error) => {
				handleError(error)
			})
		setAwaitingResponse(false)
	}

	const onConfirmLeave = async (leaveNow: boolean, leaveTime: string) => {
		setAwaitingResponse(true)

		try {
			if (!leaveTargetId) throw new Error(snackbarMesages.noValidSlot)

			// check slot for validation
			const slot = await SlotsApi.getSlot(leaveTargetId)
			const { parkedType, parkTime, vehicle, type } = _.get(slot, 'data')

			const formatLeaveTime = moment(leaveTime).format()
			if (parkTime >= formatLeaveTime) {
				throw new Error(snackbarMesages.parkTimeValidation)
			}
			// throw error if no car parked
			if (!parkedType || !vehicle) {
				throw new Error(snackbarMesages.noCarParked)
			}

			// get all transactions within 1 hour
			const getVehicleTransactionsWithin1Hour =
				await EarningsApi.getVehicleTransactionsWithin1Hour(vehicle)
			const transactions = _.get(getVehicleTransactionsWithin1Hour, 'data', [])

			// compute current fee
			const currentTimeDiff = getTimeDifference(parkTime, leaveTime)
			const currentTmeDiffRoundUp = Math.ceil(currentTimeDiff)
			const currentFee = computeTransaction(currentTmeDiffRoundUp, type)

			// check if there is transaction
			const isThereTransaction = Boolean(transactions.length)
			if (isThereTransaction) {
				// apply continous rate
				const totalFees = continousRateCalculation(
					transactions,
					currentTimeDiff,
					type,
					leaveTime
				)
				// store the exact hour not the rounded one
				await EarningsApi.add({
					price: totalFees,
					hours: currentTimeDiff,
					parkTime,
					parkingType: type,
					transactionDate: moment().format(),
					vehicle,
				})
			} else {
				// not eligible for continous rate
				await EarningsApi.add({
					price: currentFee,
					hours: currentTimeDiff,
					parkTime,
					parkingType: type,
					transactionDate: moment().format(),
					vehicle,
				})
			}

			const isLeaveNow = leaveNow || leaveTime === ''
			if (isLeaveNow) await SlotsApi.leaveSlot(leaveTargetId)
			else await SlotsApi.setLeaveSlot(leaveTargetId, formatLeaveTime)

			slotRefetch()
			earningsRefetch()
		} catch (error) {
			handleError(error)
		}

		setAwaitingResponse(false)
	}

	return {
		awaitingResponse,

		openLeave,
		openCancelModal,

		leaveTargetId,
		cancelTargetId,

		openModalReservation,
		closeModalReservation,

		openModalLeave,
		closeModalLeave,

		onConfirmCancelReservation,
		onConfirmLeave,
	}
}

export default useParkingPrompt
