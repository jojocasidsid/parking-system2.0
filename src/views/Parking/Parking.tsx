import React from 'react'
import { useSnackbar } from 'notistack'
import { useFetchSlot } from 'hooks/useFetchSlot'
import { ISlotData } from 'views/Parking/components/Slot/types'
import LoadingIndicator from 'components/LoadingIndicator'
import Page from 'components/Page'
import Slot from './components/Slot'

import { SlotsWrapper, ParkingWrapper, StyledRoot } from './styles'

const Parking = () => {
	const { enqueueSnackbar } = useSnackbar()

	const {
		isLoading: isLoadingSlot,
		data: slotData,
		refetch: slotRefetch,
	} = useFetchSlot()

	const handleLeave = (id: number) => {
		console.log(id)

		slotRefetch()

		enqueueSnackbar('Something went wrong', {
			variant: 'error',
		})
	}

	if (isLoadingSlot) {
		return <LoadingIndicator />
	}

	return (
		<Page title='Parking System'>
			<StyledRoot>
				<ParkingWrapper>
					<SlotsWrapper>
						{slotData &&
							slotData.data.map((row: ISlotData, key: number) => (
								<Slot key={key} data={row} handleLeave={(id) => handleLeave(id)} />
							))}
					</SlotsWrapper>
				</ParkingWrapper>
			</StyledRoot>
		</Page>
	)
}

export default Parking
