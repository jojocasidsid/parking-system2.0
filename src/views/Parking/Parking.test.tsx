import React from 'react'
import renderer from 'react-test-renderer'

import { QueryClient, QueryClientProvider } from 'react-query'
import { SnackbarProvider } from 'notistack'
import { ThemeProvider } from '@mui/material/styles'

import theme from 'config/theme'

import Parking from './Parking'

const queryClient = new QueryClient()

describe('Parking Page', () => {
	it('should render Parking Page without crashing', () => {
		const tree = renderer.create(
			<QueryClientProvider client={queryClient}>
				<ThemeProvider theme={theme}>
					<SnackbarProvider
						maxSnack={5}
						anchorOrigin={{
							vertical: 'bottom',
							horizontal: 'right',
						}}
						preventDuplicate
						autoHideDuration={5000}
					>
						<Parking />
					</SnackbarProvider>
				</ThemeProvider>
			</QueryClientProvider>
		)

		expect(tree).toMatchSnapshot()
	})
})
