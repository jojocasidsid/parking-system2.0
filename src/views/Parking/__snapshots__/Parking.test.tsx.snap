// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parking Page should render Parking Page without crashing 1`] = `
<div
  className="MuiBox-root css-gmuwbf"
  data-testid="loading-indicator"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-19r7jhb-MuiCircularProgress-root"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;
